CAPTCHA problem and anonymity

anuran-roy opened the pull request:
Added proxy modules (experimental) with google search results integration.

anuran-roy committed:
Added proxy.py (experimental)

The proxy script containing implementation for fetching, storing, reading and rotating proxies.

anuran-roy committed:
Delete google.py

anuran-roy committed:
Modified google.py (Experimental)

Added proxy functionalities to the Google search util class

anuran-roy committed:
Changed proxy.py (Experimental)

Made a few changes to proxy.py

anuran-roy committed:
Changed proxy.py (experimental)

Added verbose mode switch.

anuran-roy committed:
Modified initial.py

Added autoproxy option in script.

saeeddhqan commented on core/util/google.py:
```diff
@@ -17,16 +17,18 @@
 import re
 import os
-
+from core.util import proxy
```
self.framework.proxy?

anuran-roy replied:
Done.

saeeddhqan commented on core/util/google.py:
```diff
@@ -17,16 +17,18 @@
 class main:
 	# framework = None
-	def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None):
+	countip=0
```
not here

anuran-roy replied:
Sir I didn't get what you meant here. 😅

Should I remove `countip=0`? If yes, how do I replace it? 😅
It'll be really helpful if you could tell me.

saeeddhqan replied:
Define it in init

anuran-roy replied:
Done.

saeeddhqan commented on core/util/google.py:
```diff
@@ -17,16 +17,18 @@
-
+from core.util import proxy
 class main:
 	# framework = None
```
remove it.

anuran-roy replied:
Done.

rachejazz commented:
@anuran-roy the slack channels had some complaints that google is having captcha triggered always. Have you been facing any errrors?

saeeddhqan commented on core/util/google.py:
```diff
@@ -47,18 +50,29 @@ def run_crawl(self):
-				break
-
+				countip+=1
+				if(proxy.rotateip(countip)==-1):
```
if proxy.rotateip(countip) == -1:

anuran-roy replied:
Done

saeeddhqan commented on core/util/google.py:
```diff
@@ -47,18 +50,29 @@ def run_crawl(self):
-
+				countip+=1
+				if(proxy.rotateip(countip)==-1):
+					print('[PROXY] End of proxy list. ')
```
self.framework.output

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+import requests
```
copyright comment

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+
+    def getproxy(self,use=False):
+        if self.use==True:
+            print("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
```
self.output

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+    proxydict={}
+
+    def getproxy(self,use=False):
+        if self.use==True:
```
spaces: if self.use == True:

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+        '''
+        self.use=use
+        self.verbose=verbose
+    iplist=[]
```
not here

anuran-roy replied:
Sir I didn't quite get what you meant. 😅

Should I remove `iplist=[]` from here? If yes, where and how should I write it? As `self.iplist=[]`?

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+        self.use=use
+        self.verbose=verbose
+    iplist=[]
+    proxydict={}
```
not here

anuran-roy replied:
Sir I didn't quite get what you meant. 😅

Should I remove `proxydict={}` from here? If yes, where and how should I write it? As `self.proxydict={}`?

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+    def getproxy(self,use=False):
+        if self.use==True:
+            print("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
+            URL=['https://premproxy.com/list/']
```
spaces

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+            print("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
+            URL=['https://premproxy.com/list/']
+
+            i=2
```
spaces: i = 2
apply it for all

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+                        ips.append(ip)
+                    time.sleep(2)
+                #print(ips[0:5])
+                print("[PROXY] Proxies gathered.")
```
self.framework.output

anuran-roy replied:
Done

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+                #print(ips[0:5])
+                print("[PROXY] Proxies gathered.")
+            except Exception as e1:
+                print(f"An error occured. Error details: {e1}")    
```
self.framework.output for all. put a [PROXY] before the text.

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+                print(f"An error occured. Error details: {e1}")    
+            j=0    
+            while(j<len(ips)):
+                ips[j]="http://"+ips[j]
```
f"http://{ips[j]}"

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+
+            try:
+                for purl in URL:
+                    r=requests.get(purl)
```
self.framework.request? so remove requests

anuran-roy replied:
Done

anuran-roy commented:
> @anuran-roy the slack channels had some complaints that google is having captcha triggered always. Have you been facing any errrors?

@rachejazz this is the exact reason that we're working on this module. But I guess it'll take some time to be merged to the main branch, mostly because there will be quite a few changes in the existing codebase. For now, as you and others suggested, we have to switch search engines.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,108 @@
+
+class main:
+
+    def __init__(self, use=False, verbose=False):
```
use framework options. Something like this? self.framework._global_options['autoproxy')

anuran-roy replied:
Done.

saeeddhqan reviewed (commented):
most common errors: spaces, use print, request instead of self.framework.output and self.framework.request. 
Correct those and we're going to the next level.

anuran-roy committed:
Modified proxy.py and google.py

Modified the proxy and google search util classes - removed a few bugs and some imporvements

saeeddhqan commented on core/util/google.py:
```diff
@@ -17,16 +17,18 @@
 	# framework = None
-	def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None):
+	countip=0
+	def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None)#, autoproxy=False):
```
?

anuran-roy replied:
Done.

saeeddhqan commented on core/util/google.py:
```diff
@@ -37,8 +39,9 @@ def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None):
-
+		self.autoproxy=self._global_options_['autoproxy']
 	def run_crawl(self):
+		global countip
```
remove it. you have self.countip!

anuran-roy replied:
Done.

anuran-roy commented:
@saeeddhqan After your suggestions, these are the states:

[Proxy Util class](https://github.com/anuran-roy/Maryam/blob/dev-pinterest-module/core/util/proxy.py)
[Google Search Util class](https://github.com/anuran-roy/Maryam/blob/dev-pinterest-module/core/util/google.py)

I didn't get what you told me to do with the variables `countip=0`, `iplist=[]`, and `proxydict={}` 😅

saeeddhqan commented on core/util/google.py:
```diff
@@ -47,18 +50,29 @@ def run_crawl(self):
 				req = self.framework.error('[GOOGLE] Google CAPTCHA triggered.')
-				break
-
+				countip+=1
```
if autoproxy is true

anuran-roy replied:
Done.

saeeddhqan commented:
After all, is it accountable for captcha or should we use alternative search engines that use google?

anuran-roy commented:
@saeeddhqan I think that this issue will remain. This is because even if we use alternative search engines that use Google, it will be rate-limited and will have the CAPTCHA Problem again. 

What's more, the CAPTCHA appears in less crawls in those Google derivatives than the original Google search page itself. :sweat_smile:

z3y50n commented:
From my experience you can't just crawl google using proxies. They changed their policy recently and got very strict. They have an api for crawling their engine now but it requires a key and it is not free.

anuran-roy commented:
@z3y50n I tried to implement the proxy module such that it can be used for all search engines with only minimal changes in the search engine util classes. Since Google is only one among them, I'm reasonably confident that others will work with the proxy module after some slight modifications.

anuran-roy committed:
Modified google.py and proxy.py

k4u5h1k commented on core/util/google.py:
```diff
@@ -20,7 +20,6 @@
 class main:
 	# framework = None
-	countip=0
 	def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None)#, autoproxy=False):
```
Please always test your code before every push.
Missing the colon at the end.
After fixing that I got a ton of indentation errors.

anuran-roy replied:
I am really sorry for the missing colon.

But for the code indentation, I am not finding errors yet. Can you please share some screenshots here @k4u5h1k ? It'll be a great help.

k4u5h1k replied:
If it works fine push it. Could've been an error on my side.

saeeddhqan committed:
some amendments

saeeddhqan commented:
Look at the amendments and apply these code-style for the rest.

anuran-roy reviewed (commented):
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+#Python proxy module. Contributed by: @anuran-roy under @saeeddhqan
+class main:
+
+    def __init__(self, verbose = False):
```
we don't need verbose. it's usefulness so delete it comprehensively.

anuran-roy replied:
Done

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+
+    def __init__(self, verbose = False):
+        '''
+        use : Determine whether to use proxy or not.
```
remove the comment

anuran-roy replied:
Done

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+            URL = [ 'https://premproxy.com/list/' ]
+            i = 2
+            while i<=8:
+                purl = 'https://premproxy.com/list/0'+str(i)+".htm"
```
: ), purl = f"....{i}.htm"

anuran-roy replied:
Done

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+            self.framework.output("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
+            URL = [ 'https://premproxy.com/list/' ]
+            i = 2
+            while i<=8:
```
while i <= 8:

anuran-roy replied:
I guess that this is important. Premproxy has only about 8 pages. Hence the limit @saeeddhqan . 😅

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+    def getproxy(self):
+        if self.use == True:
+            self.framework.output("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
+            URL = [ 'https://premproxy.com/list/' ]
```
don't use uppercase

anuran-roy replied:
Done

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+            ips = []
+            ips2 = []
+            purl = URL[0]
+            try:
```
hard-coding. find the piece of code that yields an error, put it in the try block. rest should be in the else block.

anuran-roy replied:
I am reusing the variable, so that's why I am reinitializing them. Will save a few bytes of memory I guess 😅

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+            except Exception as e1:
+                self.framework.error(f"[PROXY] An error occured. Error details: {e1}")    
+            j = 0
+            while(j<len(ips)):
```
while j < len(ips):

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+                ips[j] = f"http://{ips[j]}"
+                j += 1
+            j = 0
+            self.framework.output("[PROXY] Now checking for availability of the proxies...")
```
simple quote: '

anuran-roy replied:
Done.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+                j += 1
+            j = 0
+            self.framework.output("[PROXY] Now checking for availability of the proxies...")
+            while(j<len(ips)):
```
while j < len(ips):

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+            j = 0
+            self.framework.output("[PROXY] Now checking for availability of the proxies...")
+            while(j<len(ips)):
+                try:
```
hard-coding

anuran-roy replied:
I reused the variable. I guess that'd save a few extra bytes of memory from being used. 😅

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+            self.framework.output("[PROXY] Now checking for availability of the proxies...")
+            while(j<len(ips)):
+                try:
+                    if int(self.framework.requests.get(ips[j]).status_code)==200:
```
bad idea. hard-coding. space errors.

anuran-roy replied:
Spacing errors corrected. 

I don't get it 😅 The HTTP status code of OK has to be hardcoded, doesn't it? Or is there any other way to implement this?

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+                        ips2.append(ips[j])
+                    if self.verbose == True:
+                        self.framework.output(f"[PROXY] {ips[j]} is available! Yay! :D")
+                except Exception as e2:
```
bad idea. use self.framework.verbose

anuran-roy replied:
Removed verbose.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+                try:
+                    if int(self.framework.requests.get(ips[j]).status_code)==200:
+                        ips2.append(ips[j])
+                    if self.verbose == True:
```
bad idea.

anuran-roy replied:
Removed verbose.

saeeddhqan commented on core/util/proxy.py:
```diff
@@ -0,0 +1,115 @@
+                    j += 1 
+                ips = []                   
+            #self.framework.output(ips2)
+            self.framework.output("[PROXY] Availability check completed! Writing to file...")
```
quote error

anuran-roy replied:
Done.

saeeddhqan reviewed (commented):
Your code style is not good. all of them need to rewrite. and I think we can just use it as a module that will find proxies?
What would you think?

anuran-roy commented:
> Your code style is not good. all of them need to rewrite. and I think we can just use it as a module that will find proxies?
> What would you think?

Yes Sir, we can just keep it as a module. I'm completely fine with that. 😄 

And as for the rewriting of code, I'm working on it. 😃

saeeddhqan commented:
Okay, thank you

anuran-roy commented:
By the way, I noticed something today. Whenever we're calling to rotateip, it isn't working because we haven't fetched IPs from online. I think that we should do something to trigger the ip fetcher method whenever someone does `set autoproxy True`. This is probably the reason why google.py isn't working.

anuran-roy reviewed (commented):
> By the way, I noticed something today. Whenever we're calling to rotateip, it isn't working because we haven't fetched IPs from online. I think that we should do something to trigger the ip fetcher method whenever someone does `set autoproxy True`. This is probably the reason why google.py isn't working.

This is regarding the modified google.py script, which doesn't fetch any results from social media. This doesn not show any error - it simply shows no results. @k4u5h1k can you please verify it? 😅

anuran-roy committed:
Made a few changes as I was told

k4u5h1k commented:
It does give an error
![Screenshot 2021-04-08 at 10 30 49 AM](https://user-images.githubusercontent.com/59250093/113970979-7eaf6c80-9855-11eb-9dc4-7c54d2f4bc07.png)   
change `self._global_options_` to `self.framework._global_options`   

After fixing that I am getting  
`[!] GoogleCodeError: getproxy() missing 1 required positional argument: 'self'.`
All the best

saeeddhqan commented on core/util/google.py:
```diff
@@ -37,7 +38,10 @@ def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None):
 		self.google_cx = google_cx
 		self._links = []
-
+		self.autoproxy = self._global_options_['autoproxy']
```
it gives error

anuran-roy commented:
Seems like this is an issue that will take quite some time to resolve. I'll try my best to resolve it as soon as possible, but it seems that it will take some time. 😞

anuran-roy commented:
I'm really sorry for my inactivity for a few days... Really really sorry for that. I was unwell.

BTW, I am stuck with a small issue... Does Maryam have a self.framework.request implementation of `requests.get`? Because when I did `self.framework.request.get`, it threw up an error.

z3y50n commented:
You use `self.framework.request` and specify the method as a parameter.

You can check how the function is defined here:

https://github.com/saeeddhqan/Maryam/blob/f938913d962b23f5090d6c88553a87eaeda29a39/core/core.py#L481-L515

anuran-roy committed:
Delete google.py

anuran-roy committed:
Create google.py

anuran-roy committed:
Merge pull request #2 from saeeddhqan/master

update

anuran-roy committed:
Added proxy.py, modified google.py

Modified google.py to work with rotating proxies.

anuran-roy commented:
Finally google.py and proxy.py are working properly. But I have a small issue. It is regarding setting autoproxy on or off. For now, I have set them to true. I have commented out what I wrote when I tried to fetch the autoproxy permissions, so you can see what I tried actually, and tell me what changes to make.

Working screenshot:

![image](https://user-images.githubusercontent.com/76481787/114730113-32d15b80-9d5e-11eb-94d3-7b8f6ab76504.png)

This is the first try. Almost immediately afterwards, I tried it again. In normal case, I would've got an error `Google CAPTCHA Triggered`

Here's the second try:

![image](https://user-images.githubusercontent.com/76481787/114730634-ad9a7680-9d5e-11eb-9fb2-27d26dba718d.png)

@saeeddhqan @k4u5h1k @rachejazz I think there are some code style inconsistencies, so I'll be working towards improving them. 😅

anuran-roy committed:
Add files via upload

saeeddhqan commented:
Check with a big doze as I told you previously.

anuran-roy commented:
Yeah sure! I'll work on them as soon as I can!

anuran-roy committed:
Add files via upload

anuran-roy committed:
Fixed Code Style inconsistencies

anuran-roy commented:
Okay, I fixed most of the coding inconsistencies I guess. Only issue remaining for now is in the environment variable I guess. Autoproxy is getting triggered even if I am not mentioning `set autoproxy True`

That's all I guess. 😅 

@saeeddhqan can you please see it?

saeeddhqan commented:
first: Check with a big doze as I told you previously.
then, show the results with estimated time.

anuran-roy commented:
After hammering Google search results non-stop, I got the trigger that `Google CAPTCHA Triggered`. But then, I get an error from the document parser that the document to parse it is NIL. Even though I gave the `continue` statement, it doesn't seem to stop the page from being loaded into the pages list.

Screenshots:

1. ![image](https://user-images.githubusercontent.com/76481787/114866245-f448a900-9e10-11eb-8043-9a477d3872d2.png)

(Error)

2.  
![image](https://user-images.githubusercontent.com/76481787/114866478-3f62bc00-9e11-11eb-87b9-3c8aec4a43d9.png)

(Concerned logical block of code)

@saeeddhqan @k4u5h1k @z3y50n @rachejazz any idea how to overcome this?

anuran-roy committed:
Some changes in `google.py` and `proxy.py`

1. Changed `google.py` so that proxies aren't initialized repeatedly.
2. Changed `proxy.py` such that availability of proxies are checked before applying them.
